frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,372 @@
+
+-- opens user dictionary
+function ReaderUserHyph:openDictionary()
+    self.dict_file = self.ui.userhyph:getDictionaryPath()
+    self.new_dict_file = self.dict_file .. ".new"
+
+    -- if no file, crate an empty one
+    if lfs.attributes(self.dict_file, "mode") ~= "file" then
+        self.dict = io.open(self.dict_file, "w")
+        self.dict:close()
+    end
+
+    -- check user dict file
+    if string.sub(lfs.attributes(self.dict_file).permissions, 1, 2) ~= "rw" then
+        UIManager:show(InfoMessage:new{
+            text = T(_("Wrong file permissions for:\n\"%1\""), self.dict_file),
+            show_icon = true,
+        })
+        return
+    end
+
+    -- open files
+    self.dict = io.open(self.dict_file, "r")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.dict_file)
+        return
+    end
+    self.new_dict = io.open(self.new_dict_file, "w")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.new_dict_file)
+        return
+    end
+    return true
+end
+
+-- reads the first dictionary entries from the old dictionary
+-- and write them to the new dict, as long the entry is alphabetically lower than word
+-- if the word matches the dictionary entry, the hypenation part is formatted and returned
+function ReaderUserHyph:findEntry(word)
+    -- scan hyphenation dictionary for selected word
+    local word_lower = Utf8:lower(word)
+    self.next_line = self.dict:read()
+    while self.next_line and Utf8:lower(self.next_line:sub(1, self.next_line:find(";") - 1)) < word_lower do
+        self.new_dict:write(self.next_line .. "\n")
+        self.next_line = self.dict:read()
+    end
+
+    local hyphenation
+    -- check if a hyphenation is found for word
+    if self.next_line and Utf8:lower(self.next_line:sub(1, self.next_line:find(";") - 1)) == word_lower then
+        hyphenation = string.sub(self.next_line, self.next_line:find(";") + 1) -- hyphenation found
+        self.next_line = nil -- Important for not duplicating the entry in the dictionary
+    end
+    return hyphenation
+end
+
+-- writes on entry to new dictionary
+function ReaderUserHyph:writeEntry(line)
+    self.new_dict:write(Utf8:lower(line) .. "\n")
+end
+
+-- reads the rest of the old dictionary and writes this to the reset
+-- if self.new_dict~=nil, an additional line can be written
+function ReaderUserHyph:writeRest()
+    -- write old entry if there was one
+    if self.next_line then
+        self.new_dict:write(self.next_line .. "\n")
+    end
+    -- copy rest of file
+    repeat
+        self.next_line = self.ui.userhyph.dict:read()
+        if self.next_line then
+            self.new_dict:write(self.ui.userhyph.next_line .. "\n")
+        end
+    until (not self.next_line)
+end
+
+-- closes all open files and invalidates variables
+-- mode
+--    KEEP_ORIGINAL .. keep the original dict
+--    USE_NEW .. use the new dict
+function ReaderUserHyph:closeDictionary(mode)
```

Comment:
Simpler to use `save_new = true/false` instead of `mode=one of these 2 constants`.